contracts/CryptoMarriage.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.7;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721BurnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+
+contract CryptoMarriage is
+  Initializable,
+  ERC721Upgradeable,
+  ERC721BurnableUpgradeable,
+  AccessControlUpgradeable
+{
+  using CountersUpgradeable for CountersUpgradeable.Counter;
+
+  struct Marriage {
+    uint256 referenceTime;
+    uint256[] dayOffsets;
+    string[] uris;
+  }
+
+  bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+  mapping(uint256 => Marriage) public marriages;
+
+  CountersUpgradeable.Counter private _tokenIdCounter;
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() initializer {}
+
+  function initialize() public initializer {
+    __ERC721_init("CryptoMarriage", "CM");
+    __ERC721Burnable_init();
+    __AccessControl_init();
+
+    _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    _grantRole(MINTER_ROLE, msg.sender);
+  }
+
+  function createMarriage(
+    address to,
+    uint256 _referenceTime,
+    string[] calldata _uris,
+    uint256[] calldata _daysOffsets
+  ) public {
+    require(
+      hasRole(MINTER_ROLE, msg.sender),
+      "Only minters can create marriages"
+    );
+    require(
+      _referenceTime < block.timestamp,
+      "Reference time must be in the past"
+    );
+    uint256 tokenId = _tokenIdCounter.current();
+    _tokenIdCounter.increment();
+    _safeMint(to, tokenId);
+
+    require(
+      _uris.length == _daysOffsets.length + 1,
```

Comment:
one option to consider to avoid this kind of constraints would be to define a new struct

```solidity
  struct TimeOffsetURI {
    uint256 dayOffset;
    string uri;
  }
```

And then your main struct becomes something like this:

```solidity
  struct Marriage {
    uint256 referenceTime;
    string referenceURI;

    TimeOffsetURI[] offsets;
  }
```

Reply:
That's much better, thanks

Reply:
Didn't manage to get around the memory struct array -> storage issue, even by copying by hand :(
I'll dig further, but I'll deploy as-is in the meantime (I think it would also take more gas ? 🤷 )